analyzers/src/SonarAnalyzer.CSharp/Rules/UseCachedRegex.cs

```diff
@@ -0,0 +1,194 @@
+    {
+        return condition switch
+        {
+            BinaryExpressionSyntax when condition.IsAnyKind(SyntaxKind.EqualsExpression) => CheckCorrectBranchResult(conditional.WhenTrue, conditional.WhenFalse),
+            BinaryExpressionSyntax when condition.IsAnyKind(SyntaxKind.NotEqualsExpression) => CheckCorrectBranchResult(conditional.WhenFalse, conditional.WhenTrue),
+            { RawKind: (int)SyntaxKindEx.IsPatternExpression } =>
+                (IsPatternExpressionSyntaxWrapper)condition switch
+                {
+                    { Pattern.SyntaxNode.RawKind: (int)SyntaxKindEx.ConstantPattern } => CheckCorrectBranchResult(conditional.WhenTrue, conditional.WhenFalse),
+                    { Pattern.SyntaxNode.RawKind: (int)SyntaxKindEx.NotPattern } => CheckCorrectBranchResult(conditional.WhenFalse, conditional.WhenTrue),
+                    _ => false
+                },
+            _ => false
+        };
+
+        bool CheckCorrectBranchResult(ExpressionSyntax first, ExpressionSyntax second) =>
+            first is ObjectCreationExpressionSyntax && second.GetName() == assignment.Left.GetName();
+    }
+
+    private static bool IsConditionInvolveAssigned(ExpressionSyntax condition, AssignmentExpressionSyntax assignment) =>
+        condition switch
+        {
+            BinaryExpressionSyntax binaryCondition when condition.IsAnyKind(SyntaxKind.EqualsExpression, SyntaxKind.NotEqualsExpression) =>
+                ChecksForNullOnAssigned(assignment, binaryCondition.Left, binaryCondition.Right) || ChecksForNullOnAssigned(assignment, binaryCondition.Right, binaryCondition.Left),
+            { RawKind: (int)SyntaxKindEx.IsPatternExpression } =>
+                (IsPatternExpressionSyntaxWrapper)condition switch
+                {
+                    { Expression: { } expression, Pattern: { SyntaxNode.RawKind: (int)SyntaxKindEx.ConstantPattern } pattern } =>
+                        ChecksForNullOnAssigned(assignment, expression, ((ConstantPatternSyntaxWrapper)pattern).Expression),
+                    { Expression: { } expression, Pattern: { SyntaxNode.RawKind: (int)SyntaxKindEx.NotPattern } notPattern } =>
+                        (UnaryPatternSyntaxWrapper)notPattern is { Pattern: { SyntaxNode.RawKind: (int)SyntaxKindEx.ConstantPattern } pattern }
+                        && ChecksForNullOnAssigned(assignment, expression, ((ConstantPatternSyntaxWrapper)pattern).Expression),
+                    _ => false
+                },
+            _ => false
+        };
+
+    private static bool IsAssignmentWithinIfStatement(AssignmentExpressionSyntax assignment, SemanticModel model)
+    {
+        SyntaxNode node = assignment;
+        while (!node.Parent.IsKind(SyntaxKind.CompilationUnit))
+        {
+            if (node.Parent.IsKind(SyntaxKind.IfStatement))
+            {
+                break;
+            }
+
+            node = node.Parent;
+        }
```

Comment:
You can have a separate method for this to remove the duplicated functionality with IsWithinCorrectContext